Verify with checksum when signature not available

dbnicholson opened the pull request:
This needs a little polish but I wanted to get some feedback. If there's no `.asc` file next to the image, try looking for a `.sha256` file and do a checksum verification.

https://phabricator.endlessm.com/T31304

dbnicholson committed:
scribe: Allow verifying image with sha256 checksum

As an alternative to requiring a GPG signature from a trusted key, allow
verifying the image with a sha256 checksum. The checksum file is
expected to reside alongside the image with a `.sha256` suffix.
Typically this would be a file formatted by the `sha256sum` program, but
any file where the first word is a sha256 hex digest is acceptable.

While this seems less secure, in reality it provides no more validation
than using GPG since the trusted key is embedded in the payload. An
attacker could alter the signature and keyring to make it appear that it
came from the expected party. So, all that the signature or checksum can
truly validate is that the image file contains the bytes the person
providing it expected.

https://phabricator.endlessm.com/T31304

dbnicholson commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -1150,6 +1186,128 @@ gis_scribe_begin_verify (GisScribe *self,
+  if (!g_file_load_contents (self->checksum, cancellable, &checksum_contents,
+                             &checksum_len, NULL, &error))
+    {
+      task_return_error (self, task, g_steal_pointer (&error));
```
I think this is an existing bug, but currently this will deadlock the installer using `task_return_error` because the `gis_scribe_begin_*` functions used for task setup are called with the mutex locked. I briefly changed them back to `g_task_return_error`, but then the error message that shows up is wrong because it doesn't set `GisScribe::error`. I'm not sure what the best approach is.

wjt replied:
Ah, I'd forgotten this particular wart. I was never too happy with how the subtasks' errors propagate upwards. I'm not sure if there is more advanced technology for multiplexed GTask these days.

Perhaps the different fields could be guarded by different mutexes? Looking at the code, I think `gis_scribe_write_thread_await_gpg()` could be made to only access `self->outstanding_tasks`…

dbnicholson replied:
You could also change the `outstanding_tasks` usage to atomic integer access, but I'm not sure that's the right model since the top level seems like it wants to keep the lock until all the tasks are setup.

wjt replied:
Oh, that could work. I think I wanted to have the property that it would always be non-zero at the end of this function, so that one of the earlier subtasks failing couldn't cause the task we're setting up to be completed while we're still going.

I don't actually know if that's possible. And it could be faked by adding an extra field to the bitfield for "SETUP".

dbnicholson replied:
I ended up just locking and unlocking the mutex around each access to `outstanding_tasks` in `gis_scribe_write_async`. Initially I thought it was important that the mutex was held until all the subtasks were started, but I don't think that's necessary. I added tests for this (and manually tested by running the installer with no signature present) and it did the right thing.

wjt replied:
OK, makes sense!

wjt commented on gnome-image-installer/pages/diskimage/gis-diskimage-page.c:
```diff
@@ -117,6 +119,12 @@ gis_diskimage_page_selection_changed(GtkWidget *combo, GisPage *page)
+    checksum = g_strjoin (NULL, image, ".sha256", NULL);
+
+  gis_store_set_image_checksum (checksum);
+  g_free (checksum);
```
These days I would declare this variable as:

```
g_autofree char *checksum = NULL;
```

and then not need the manual `g_free`.

dbnicholson replied:
Yes, that would be better. `signature` above is doing the same thing.

wjt commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -1507,11 +1667,32 @@ gis_scribe_write_async (GisScribe          *self,
+          task, GIS_IMAGE_ERROR, GIS_IMAGE_ERROR_VERIFICATION_FAILED,
+          _("Neither the signature file ‘%s’ nor the checksum file ‘%s’ exist."),
+          signature_path, checksum_path);
+      goto out;
```
Won't this skip the block just below that closes the other pipes?

dbnicholson replied:
Oh, good point. I don't think I fully understand the error handling here.

wjt commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -1150,6 +1186,128 @@ gis_scribe_begin_verify (GisScribe *self,
+      return NULL;
+    }
+
+  if (checksum_len < CHECKSUM_STRLEN)
+    {
+      task_return_new_error (
+          self, task, GIS_IMAGE_ERROR, GIS_IMAGE_ERROR_VERIFICATION_FAILED,
+          _("The checksum file ‘%s’ does not contain %d characters."),
+          checksum_path, CHECKSUM_STRLEN);
+      return NULL;
+    }
```
Seems redundant with the block below?

dbnicholson replied:
Right. I had something different originally but this is unnecessary.

wjt commented on gnome-image-installer/pages/install/gis-scribe.c:
```diff
@@ -1150,6 +1186,128 @@ gis_scribe_begin_verify (GisScribe *self,
+        task_return_new_error (
+            self, task, GIS_IMAGE_ERROR, GIS_IMAGE_ERROR_VERIFICATION_FAILED,
+            _("The checksum file ‘%s’ contains invalid character ‘%c’."),
+            checksum_path, *cur);
```
`*cur` might be part of a multi-byte UTF-8 sequence. I wonder what would be substituted into the string in that case?

dbnicholson replied:
Garbage? I didn't try to handle this, but let me know what you think.

wjt replied:
It's probably fine as-is.

wjt reviewed (commented):
Had a couple of superficial comments but structurally this seems fine. Next stop, https://github.com/endlessm/eos-image-builder/pull/2.

dbnicholson committed:
scribe: Fix deadlock from async initialization errors

If there is an error setting up any of the tasks in
`gis_scribe_write_async`, it needs to be propagated to
`GisScribe::error`. This requires locking the mutex, but currently
that's already locked throughout `gis_scribe_write_async` to guard
changes to `GisScribe::outstanding_tasks`. This causes a deadlock if the
various `gis_scribe_begin_*` setup functions try to set an error since
they won't return.

A concrete example of this is if an image signature file doesn't exist.
The installer will instead show no progress until it's killed. Instead,
lock and unlock the mutex around each access to
`GisScribe::outstanding_tasks`. Another option would be to use a
separate locking mechanism for the outstanding tasks and error, but this
seemed simpler.

https://phabricator.endlessm.com/T31304

dbnicholson committed:
tests: Add tests for missing scribe components

These cause `gis_scribe_write_async` to fail before the write task has
been setup, which means the zeroing check needs to be skipped. Prior to
the previous commit this would cause a deadlock and the tests would
eventually timeout.

https://phabricator.endlessm.com/T31304

dbnicholson commented:
OK, I think this is ready now. The first commit is a bug fix (with tests in the 2nd commit) which I can split out to a separate PR if you prefer.

wjt reviewed (approved):
Looks great, thanks!
